kong/resty/dns_client/init.lua

```diff
@@ -0,0 +1,750 @@
+      elseif answer.type == TYPE_SRV then
+        answer.target = ipv6_bracket(answer.target)
+      end
+
+      -- skip the SRV record pointing to itself,
+      -- see https://github.com/Kong/lua-resty-dns-client/pull/3
+      if not (answer.type == TYPE_SRV and answer.target == qname) then
+        table_insert(processed_answers, answer)
+      end
+    end
+
+    if self.valid_ttl then
+      answer.ttl = self.valid_ttl
+    else
+      ttl = math_min(ttl, answer.ttl)
+    end
+  end
+
+  if #processed_answers == 0 then
+    if not cname_answer then
+      return {
+        errcode = EMPTY_RECORD_EC,
+        errstr  = EMPTY_RECORD_ESTR,
+        ttl     = self.empty_ttl,
+        -- expire = now() + self.empty_ttl,
+      }
+    end
+
+    table_insert(processed_answers, cname_answer)
+  end
+
+  processed_answers.expire = now() + ttl
+  processed_answers.ttl = ttl
+
+  return processed_answers
+end
+
+
+local function resolve_query(self, name, qtype, tries)
+  local key = name .. ":" .. qtype
+  stats_count(self.stats, key, "query")
+
+  local r, err = resolver:new(self.r_opts)
+  if not r then
+    return nil, "failed to instantiate the resolver: " .. err
+  end
+
+  local options = { additional_section = true, qtype = qtype }
+  local answers, err = r:query(name, options)
+  if r.destroy then
+    r:destroy()
+  end
+
+  if not answers then
+    stats_count(self.stats, key, "query_fail_nameserver")
+    return nil, "DNS server error: " .. (err or "unknown")
+  end
+
+  answers = process_answers(self, name, qtype, answers)
+
+  stats_count(self.stats, key, answers.errstr and "query_fail:" .. answers.errstr
+                                               or "query_succ")
+
+  return answers, nil, answers.ttl
+end
+
+
+local function stale_update_task(premature, self, key, name, qtype, short_key, ttl)
+  if premature then
+    return
+  end
+
+  local answers = resolve_query(self, name, qtype, {})
+  if not answers then
+    -- retry update after failure
+    local retry_delay = math_random(STALE_UPDATE_DELAY, STALE_UPDATE_DELAY * 2)
+    ttl = ttl - retry_delay
+    if ttl < 0 then
+      return  -- no need to retry if it exceeds the stale_ttl
+    end
+    local ok, err = timer_at(retry_delay, stale_update_task, self, key, name,
+                             qtype, short_key, ttl)
+    if not ok then
+      log(ALERT, "failed to start a timer to re-update stale DNS records: ", err)
+    end
+    return
+  end
+
+  if not answers.errcode or answers.errcode == NAME_ERROR_CODE then
+    self.cache:set(key, { ttl = answers.ttl }, answers)
+    insert_last_type(self.cache, name, qtype)
+
+    -- simply invalidate it and let the search iteration choose the correct one
+    self.cache:delete(short_key)
+  end
+end
+
+
+local function start_stale_update_task(self, key, name, qtype, short_key, ttl)
+  stats_count(self.stats, key, "stale")
+
+  local ok, err = timer_at(0, stale_update_task, self, key, name, qtype,
```

Comment:
I might be missing something but how is the concurrency control in this call done?

Reply:
As we talked private, the concurrency control is implemented by the resty.lock before mlcache run L3 callback function.

Reply:
And I'll include a comment here for other readers.

Reply:
done in this commit https://github.com/Kong/kong/pull/12305/commits/7070bf3d22f8b27e8bd94b3c367ed110b95fbb61